### ISSUE TITLE
Cert authentication in Postgres-XC

### ISSUE THREAD

**Issue by dimv36** [postgres-x2/postgres-x2#295]
I have a problem with cert authentication in Postgres-XC.

Settings, that were changed:
1. I create server's certificate server.crt and public key, signed by CA.
2. I create user's certificate, signed by CA.
3. Config files:
in pg_hba.conf:
hostssl all all 127.0.0.1/32 cert clientcert=1

in postgresql.conf:
ssl = on
ssl_cert_file = 'server.crt' 
ssl_key_file = 'server.key'
ssl_ca_file = 'root.crt'
at coordinator and datanodes.

So, I restart clusters, and connect by postgres-xc to coordinator. I try to create table:
CREATE TABLE test(id integer);
but I have message: Failed to get pooled connections.

Can you help me?

**Comment by koichi-szk** [comment]
Hmm, there could be some issues in coordinator-datanode connections.   Did
you do the same settings in the datanodes?

Unfortunately, I don't have an experience to test SSL connection between
coordinator and datanode.

Could you test the connection itself before creating tables?  I mean,
connect to your coordinator via psql and run some statements such as select
- from pg_class.   If it is successful, could you run the same statement
  through EXECUTE DIRECT on (datanode)?

If it is successful, there could be some authentication problem.   If not,
problem will be in SSL link.

Thank you very much;

---

Koichi Suzuki
https://www.linkedin.com/in/koichidbms

2015-09-09 16:11 GMT+09:00 Dmitry Voronin notifications@github.com:

**Comment by dimv36** [comment]
Yes, I have the same settings on all nodes (coordinator and datanodes).

I connect to coordinator, datanode1 and datanode2 and execute query: SELECT \* FROM pg_class;
It works. Next, I connect to coordinator and execute query EXECUTE DIRECT ON (dn1) 'SELECT \* FROM pg_class;'; and get error message, that I write above.

So, I think, we have a problem with SSL link.

**Comment by koichi-szk** [comment]
Year, I suspect the problem is in coordinator-datanode SSL link.
Unfortunately, I don't have good experience on libpq SSL link.
Unfortunately, I'm afraid pooler is not equipped with a feature to send
passphrase or any authentication information related to SSL.

Could you disable all the SSL links between coordinator and datanode and
test the same thing?   I suspect current pooler is too simple at this point.

If you have any similar problem, please set LOG_MIN_MESSGES at each node to
lowest level (DEBUG5 or so).  Pooler may write some useful information to
see what's going on.

Best and thank you very much;

---

Koichi Suzuki
https://www.linkedin.com/in/koichidbms

2015-09-09 16:39 GMT+09:00 Dmitry Voronin notifications@github.com:

**Comment by dimv36** [comment]
When I set md5 authentication, it was Ok.

I tests SSL with settings, as above.
I set DEBUG5 but just have message:

can not connect to node 16384
failed to acquire connections
Failed to get pooled connections

On coordinator node.

On datanode:
SSL-connection from "(anonymous)"
FATAL:  connection requires a valid client certificate

I think it's problem. We create connections after starting Postgres-XC daemon, and reuse them after real starting session by user. At cert mode, we must create connections to datanodes after successful session initialization at coordinator node, maybe.

**Comment by koichi-szk** [comment]
There may be some GUC to control when the pooler acquire datanode
connection and not to reuse the connection.   I hope I have a time to help
tomorrow.

---

Koichi Suzuki
2015/09/09 17:43 "Dmitry Voronin" notifications@github.com:

**Comment by koichi-szk** [comment]
Hello;

As much as I looked into, there's no specific API for libpq client (pooler
depends on libpq) to use SSL, except for PQgetssl(),
PQinitOpenSSL() or PQinitSSL(), which are all optional.

So the next is if configuration of SSL resources required as client is set
to home directory of the linux user running pooler.   If it is okay, I
mean, if you can connect to other coordinator successfully through SSL from
a host such coordinator is running, then we should consider another cause.

As you may know, pooler assigns libpq socket using SCM_RIGHTS message.   I
suspect that we need to do something extra to copy SSL socket from the
pooler to the coordinator backend and let libpq client work correctly.

May need some more analysis on this.

Thank you;

---

Koichi Suzuki
https://www.linkedin.com/in/koichidbms

2015-09-09 18:37 GMT+09:00 Koichi Suzuki koichi.dbms@gmail.com:

**Comment by dimv36** [comment]
OK, If I have free time, I will write test code to make support for SSL connections.
